[SYCL][ClangLinkerWrapper] Add SYCL Module split library usage in clang-linker-wrapper

maksimsab opened the pull request:
This patch introduces -sycl-module-split-mode command line option in clang-linker-wrapper tool and makes the tool invoke either library or sycl-post-link tool depending on the presense of -sycl-module-split-mode option.

Also the patch removes funcionality of reading/writing of Table files in clang-linker-wrapper. Instead of them the tool starts using in-memory structures.

maksimsab commented:
Note for reviewers.
This patch functionally depends on https://github.com/intel/llvm/pull/13282.  It is supposed that the current patch will be merged strictly after https://github.com/intel/llvm/pull/13282. However, in order to make it work the corresponding functionality was copy-pasted from https://github.com/intel/llvm/pull/13282. If you have any issues about `splitting` part then, please, address them directly to https://github.com/intel/llvm/pull/13282.

maksimsab commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -624,19 +596,160 @@ struct Table {
+  }
+
+  if (Args.hasArg(OPT_print_split_modules))
+    cantFail(printSplitModules(SplitModules));
```
`ExitOnError` is better

maksimsab replied:
Not applicable.

asudarsa commented on clang/tools/clang-linker-wrapper/LinkerWrapperOpts.td:
```diff
@@ -163,6 +163,15 @@ def sycl_post_link_options_EQ : Joined<["--", "-"], "sycl-post-link-options=">,
   Flags<[WrapperOnlyOption]>,
   HelpText<"Options that will control sycl-post-link step">;
 
+def sycl_module_split_mode_EQ :
```
I think we will need to add tests to make sure that the existing tests which use the sycl-post-link will run the same way when we use the split library.

Thanks

asudarsa commented on llvm/lib/SYCLLowerIR/ModuleSplitter.cpp:
```diff
@@ -1113,5 +1138,63 @@ SmallVector<ModuleDesc, 2> splitByESIMD(ModuleDesc &&MD,
+  ModuleDesc MD = std::move(M); // makeModuleDesc() ?
+  // FIXME: false arguments are temporary for now.
+  auto Splitter =
+      getDeviceCodeSplitter(std::move(MD), Settings.Mode, false, false);
```
We can try and pass the entries from -sycl-post-link-options via the Settings variable. I am ok to do it in a subsequent PR. There will be other options like -emit-exported-symbols, -emit-imported-symbols if/when we want to add dynamic linking support here.

Thanks

maksimsab replied:
The issue here is that the library is not complete. The plan is to fulfill missing functionality and add the corresponding splitting option parsing.

asudarsa commented on llvm/lib/SYCLLowerIR/ModuleSplitter.cpp:
```diff
@@ -1113,5 +1138,63 @@ SmallVector<ModuleDesc, 2> splitByESIMD(ModuleDesc &&MD,
+  std::vector<SplitModule> OutputImages;
+  while (Splitter->hasMoreSplits()) {
+    ModuleDesc MD2 = Splitter->nextSplit();
+    MD2.fixupLinkageOfDirectInvokeSimdTargets();
```
We will not be upstreaming this. Do we want to add a TODO marker for tracking that?

Thanks

asudarsa commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -624,19 +596,160 @@ struct Table {
+}
+
+// Run sycl-post-link tool
+static Expected<std::vector<module_split::SplitModule>>
```
I suppose we will remove this path in subsequent PRs?

maksimsab replied:
Yes.

asudarsa commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -624,19 +596,160 @@ struct Table {
+}
+
+static Expected<std::vector<module_split::SplitModule>>
+runSYCLSplitLibrary(ArrayRef<StringRef> InputFiles, const ArgList &Args,
```
We will want to parse the -sycl-post-link-options here.

asudarsa replied:
Can you please share your thoughts on this? Thanks

maksimsab replied:
The issue here is that the library is not complete. The plan is to fulfill missing functionality and add the corresponding splitting option parsing.

asudarsa commented on clang/test/Driver/linker-wrapper-sycl.cpp:
```diff
@@ -7,6 +7,20 @@
+// Check sycl-module-split-mode command line option
+// RUN: clang-linker-wrapper -sycl-module-split-mode=auto -sycl-device-library-location=%S/Inputs -sycl-device-libraries=libsycl-crt.o,libsycl-complex.o -llvm-spirv-options="LLVM_SPIRV_OPTIONS" "--host-triple=x86_64-unknown-linux-gnu" "--triple=spir64" "--linker-path=/usr/bin/ld" "--" HOST_LINKER_FLAGS "-dynamic-linker" HOST_DYN_LIB "-o" "a.out" HOST_LIB_PATH HOST_STAT_LIB %S/Inputs/test-sycl.o - ...
+// CHK-SPLIT-CMDS: "{{.*}}llvm-link" [[INPUT:.*]].bc -o [[FIRSTLLVMLINKOUT:.*]].bc --suppress-warnings
+// CHK-SPLIT-CMDS-NEXT: "{{.*}}clang-offload-bundler" -type=o -targets=sycl-spir64-unknown-unknown -input={{.*}}libsycl-crt.o -output=[[FIRSTUNBUNDLEDLIB:.*]].bc -unbundle -allow-missing-bundles
```
We do not need to 'unbundle' the device library files anymore. I updated the device library file creation to create 'packaged' files instead which get 'unpackaged' in the clang-linker-wrapper.

Thanks

asudarsa reviewed (commented):
The change looks great. Specifically, the move from using 'table' file to using SplitModules is a very good segway to start using the generic 'Image' data structure (extended for SYCL).
I have added a few comments to be addressed. No major concerns from me.

Thanks

maksimsab committed:
[SYCL][ClangLinkerWrapper] Add SYCL split library usage to clang-linker-wrapper

This patch introduces -sycl-module-split-mode command line option in
clang-linker-wrapper tool and makes the tool invoke either library or
sycl-post-link tool depending on the presense of -sycl-module-split-mode
option.

Also the patch removes funcionality of reading/writing of Table files in
clang-linker-wrapper. Instead of them the tool starts using in-memory
structures.

maksimsab commented:
I will do final cleanup and reopen this PR for CR.

maksimsab committed:
clean up code and clang-format

asudarsa commented on clang/test/Driver/linker-wrapper-sycl.cpp:
```diff
@@ -15,14 +15,26 @@
 // CHK-CMDS-NEXT: "{{.*}}llc" -filetype=obj -o [[LLCOUT:.*]] [[WRAPPEROUT]].bc
 // CHK-CMDS-NEXT: "{{.*}}/ld" -- HOST_LINKER_FLAGS -dynamic-linker HOST_DYN_LIB -o a.out [[LLCOUT]] HOST_LIB_PATH HOST_STAT_LIB {{.*}}.o
 
+// Check sycl-module-split-mode command line option
```
Can you please add a line describing what happens with and without this option?

Thanks

asudarsa commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -528,6 +531,64 @@ static Expected<StringRef> convertSPIRVToIR(StringRef Filename,
   return *TempFileOrErr;
 }
 
+/// Parses the output table file from sycl-post-link tool.
```
Will it be possible to move this to ModuleSplitter? Thanks

asudarsa commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -626,79 +687,94 @@ static Expected<StringRef> runSYCLPostLink(ArrayRef<StringRef> InputFiles,
-    std::string PropFile;
-    std::string SymFile;
-  };
+  if (DryRun) {
```
Can you please add a comment here? Thanks

maksimsab replied:
Added.

asudarsa commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -626,79 +687,94 @@ static Expected<StringRef> runSYCLPostLink(ArrayRef<StringRef> InputFiles,
-    SmallVector<std::string, 16> Files;
-    for (auto &Entry : Entries) {
-      Files.push_back(Entry.IRFile);
+Error printSplitModules(const std::vector<module_split::SplitModule> &Modules) {
```
Can we move this also to ModuleSplitter? Thanks

asudarsa commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -626,79 +687,94 @@ static Expected<StringRef> runSYCLPostLink(ArrayRef<StringRef> InputFiles,
+  return Error::success();
+}
+
+static Expected<std::vector<module_split::SplitModule>>
```
Please add a comment describing this function (similar to runSYCLPostLinkTool).

maksimsab replied:
Added.

asudarsa commented:
Hi @maksimsab 
This PR looks good as an incremental change. It is a little difficult to understand the status of these changes though. In the PR description, can you please add the list of sycl-post-link features that are already available in the library (for e.g. code splitting) and also the list of sycl-post-link features that are not yet available (e.g. prop/sym handling and -sycl-post-link-options handling) and also a short plan on when/how they will be added.
In terms of testing, is my understanding correct that the library flow is not yet ready for E2E testing?

Thanks

maksimsab commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -528,6 +531,64 @@ static Expected<StringRef> convertSPIRVToIR(StringRef Filename,
+static Expected<std::vector<module_split::SplitModule>>
+parseSplitModulesFromFile(StringRef File) {
+  auto EntriesMBOrErr = llvm::MemoryBuffer::getFile(File);
+
```
Exactly. In the end I see it significantly simplified like a call `splitAndWrap()` or something like that..

sarnex replied:
sounds good, thx

maksimsab commented on clang/tools/clang-linker-wrapper/LinkerWrapperOpts.td:
```diff
@@ -169,6 +169,15 @@ def sycl_post_link_options_EQ : Joined<["--", "-"], "sycl-post-link-options=">,
+  HelpText<"Option that turns on split library with the given split mode">;
+
+def print_split_modules : Flag<["--"], "print-split-modules">,
+  Flags<[WrapperOnlyOption]>,
```
Maybe. Also this option might be removed since we have `-save-temps`.

maksimsab committed:
resolve CR issues

sarnex commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -626,79 +629,85 @@ static Expected<StringRef> runSYCLPostLink(ArrayRef<StringRef> InputFiles,
-    return *TempFileOrErr;
+  llvm::module_split::ModuleSplitterSettings Settings;
+  Settings.Mode = Mode;
+  Settings.OutputPrefix = "";
```
Just a question, are there any cases where we're going to set a prefix in the future?

maksimsab replied:
In fact, it should be changed when `-save-temps` is used.

sarnex commented on llvm/include/llvm/SYCLLowerIR/ModuleSplitter.h:
```diff
@@ -45,6 +46,10 @@ enum IRSplitMode {
   SPLIT_NONE        // no splitting
 };
 
+// \returns IRSplitMode value if \p S is recognized. Otherwise, std::nullopt is
```
nit: mix between `//` comments here and `///` for `parseSplitModulesFromFile`

sarnex reviewed (approved):
lgtm, thanks! only minor comments

mdtoguchi commented on clang/test/Driver/linker-wrapper-sycl.cpp:
```diff
@@ -90,7 +103,7 @@
 // RUN: clang-linker-wrapper -sycl-device-libraries=%t3.devicelib.o -sycl-post-link-options="SYCL_POST_LINK_OPTIONS" -llvm-spirv-options="LLVM_SPIRV_OPTIONS" "--host-triple=x86_64-unknown-linux-gnu" "--linker-path=/usr/bin/ld" "--" HOST_LINKER_FLAGS "-dynamic-linker" HOST_DYN_LIB "-o" "a.out" HOST_LIB_PATH HOST_STAT_LIB %t3.o --dry-run 2>&1 | FileCheck -check-prefix=CHK-CMDS-AOT-NV %s
 // CHK-CMDS-AOT-NV: "{{.*}}spirv-to-ir-wrapper" {{.*}} -o [[FIRSTLLVMLINKIN:.*]].bc --llvm-spirv-opts --spirv-preserve-auxdata --spirv-target-env=SPV-IR --spirv-builtin-format=global
-// CHK-CMDS-AOT-NV-NEXT: "{{.*}}llvm-link" [[FIRSTLLVMLINKIN:.*]].bc -o [[FIRSTLLVMLINKOUT:.*]].bc --suppress-warnings
+// CHK-CMDS-AOT-NV-NEXT: "{{.*}}llvm-link" [[FIRSTLLVMLINKIN]].bc -o [[FIRSTLLVMLINKOUT:.*]].bc --suppress-warnings
```
Eagle eyes to catch the re-setting of the `FIRSTLLVMLINKIN` var!

asudarsa commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -699,28 +642,17 @@ runSYCLPostLinkTool(ArrayRef<StringRef> InputFiles, const ArgList &Args) {
-  return parseSplitModulesFromFile(*TempFileOrErr);
-}
-
-Error printSplitModules(const std::vector<module_split::SplitModule> &Modules) {
```
Where has this been moved to? I do not see it in the ModuleSplitter. Can you please check? Thanks

asudarsa reviewed (commented):
LGTM. I have one outstanding question about how you plan to handle -sycl-post-link-options in the library version. I am ok with a plan to support it in a future PR. But I would like to see some path forward here. 
Thanks again for this PR.

sarnex commented:
@asudarsa Is it okay if I merge this PR? It doesn't seem like your comment is blocking. I'm dependent on this for thinLTO work. Thanks.

sarnex commented:
Sorry Arvind is on vacation for a while.

@maksimsab Is it okay if I merge this now and you can address any feedabck from Arvind when he comes back? Thanks.

maksimsab commented:
@sarnex Sure! I have the upcoming patch that depends on this one.

sarnex commented:
Thanks!
